File: kotlin/rayfold-core/src/main/kotlin/dev/rayfold/core/Expr.kt
package dev.rayfold.core

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonPrimitive
import java.math.BigDecimal

/** Policy expression evaluation over the IR's JSON AST (spec/01 section 5). Missing paths are null; ordering values that do not order throws [ExprError]. */
class ExprEnv(val viewer: JsonElement, val args: JsonElement, val self: JsonElement, val now: () -> Long = System::currentTimeMillis)

/** An expression that cannot be evaluated, such as `"high" > 3`; [Policy] fails closed on it. */
class ExprError(message: String) : RuntimeException(message)

object Expr {
    fun eval(e: JsonObject, env: ExprEnv): JsonElement {
        return when (e["k"]?.jsonPrimitive?.content) {
            "lit" -> e["v"] ?: JsonNull
            "path" -> {
                var cur: JsonElement = when (e["root"]?.jsonPrimitive?.content) {
                    "viewer" -> env.viewer
                    "args" -> env.args
                    else -> env.self
                }
                for (p in (e["path"] as JsonArray)) {
                    val o = cur as? JsonObject ?: return JsonNull
                    cur = o[p.jsonPrimitive.content] ?: return JsonNull
                }
                cur
            }
            "not" -> JsonPrimitive(!truthy(eval(e["e"] as JsonObject, env)))
            "list" -> JsonArray((e["items"] as JsonArray).map { eval(it as JsonObject, env) })
            "call" -> {
                val a = (e["args"] as JsonArray).map { eval(it as JsonObject, env) }
                when (e["fn"]?.jsonPrimitive?.content) {
                    "has" -> JsonPrimitive((a.getOrNull(0) as? JsonArray)?.any { looseEq(it, a.getOrNull(1) ?: JsonNull) } == true)
                    "len" -> when (val x = a.getOrNull(0)) {
                        is JsonArray -> JsonPrimitive(x.size)
                        is JsonPrimitive -> if (x.isString) JsonPrimitive(x.content.length) else JsonNull
                        else -> JsonNull
                    }
                    "now" -> JsonPrimitive(env.now())
                    else -> JsonNull
                }
            }
            "bin" -> {
                val op = e["op"]?.jsonPrimitive?.content
                if (op == "&&") return JsonPrimitive(truthy(eval(e["l"] as JsonObject, env)) && truthy(eval(e["r"] as JsonObject, env)))
                if (op == "||") return JsonPrimitive(truthy(eval(e["l"] as JsonObject, env)) || truthy(eval(e["r"] as JsonObject, env)))
                val l = eval(e["l"] as JsonObject, env)
                val r = eval(e["r"] as JsonObject, env)
                when (op) {
                    "==" -> JsonPrimitive(looseEq(l, r))
                    "!=" -> JsonPrimitive(!looseEq(l, r))
                    "in" -> JsonPrimitive((r as? JsonArray)?.any { looseEq(l, it) } == true)
                    else -> JsonPrimitive(order(l, r)?.let { cmp(op, it) } ?: false)
                }
            }
            else -> JsonNull
        }
    }

    private fun cmp(op: String?, c: Int): Boolean = when (op) {
        "<" -> c < 0; "<=" -> c <= 0; ">" -> c > 0; ">=" -> c >= 0; else -> false
    }

    fun truthy(v: JsonElement): Boolean = when (v) {
        is JsonNull -> false
        is JsonPrimitive -> !(v.content == "false" && !v.isString)
        else -> true
    }

    /**
     * null == null only. A number equals a number or a numeric string of the same exact value (Decimal and Long travel
     * as strings, ids may come as numbers); other scalars compare by text.
     */
    fun looseEq(a: JsonElement, b: JsonElement): Boolean {
        if (a is JsonNull) return b is JsonNull
        if (b is JsonNull) return false
        if (a !is JsonPrimitive || b !is JsonPrimitive) return false
        if (!a.isString || !b.isString) {
            val an = decimal(a)
            val bn = decimal(b)
            if (an != null && bn != null) return an.compareTo(bn) == 0
        }
        return a.content == b.content
    }

    /** Plain decimal text, as Decimal and Long are written on the wire (the same rule as the TypeScript runtime). */
    private val NUMERIC_STRING = Regex("-?[0-9]+(\\.[0-9]+)?")

    /**
     * Sign of l - r, or null when either side is null (every comparison with null is false). Numbers and numeric
     * strings order by exact value, other strings lexically; anything else does not order and throws [ExprError].
     */
    private fun order(l: JsonElement, r: JsonElement): Int? {
        if (l is JsonNull || r is JsonNull) return null
        val a = l as? JsonPrimitive
        val b = r as? JsonPrimitive
        if (a != null && b != null) {
            // Decimal and Long travel as text, so two numeric strings order by value too ("900" < "1000").
            val an = decimal(a)
            val bn = decimal(b)
            if (an != null && bn != null) return an.compareTo(bn)
            if (a.isString && b.isString) return a.content.compareTo(b.content)
        }
        throw ExprError("cannot order ${describe(l)} and ${describe(r)}")
    }

    /** A JSON number, or a string holding one, as an exact decimal (no rounding past 2^53); null for anything else. */
    private fun decimal(p: JsonPrimitive): BigDecimal? {
        val numeric = if (p.isString) NUMERIC_STRING.matches(p.content) else StrictJson.isNumber(p.content)
        return if (numeric) p.content.toBigDecimalOrNull() else null
    }

    private fun describe(v: JsonElement): String = when (v) {
        is JsonNull -> "null"
        is JsonArray -> "a list"
        is JsonObject -> "an object"
        is JsonPrimitive -> if (v.isString) "a string" else if (v.content == "true" || v.content == "false") "a boolean" else "a number"
    }

    fun referencesViewer(e: JsonObject): Boolean = paths(e).any { it == "viewer" }

    fun paths(e: JsonObject, out: MutableList<String> = mutableListOf()): List<String> {
        when (e["k"]?.jsonPrimitive?.content) {
            "path" -> out.add((e["root"] ?: error("path expression without a root")).jsonPrimitive.content)
            "bin" -> { paths(e["l"] as JsonObject, out); paths(e["r"] as JsonObject, out) }
            "not" -> paths(e["e"] as JsonObject, out)
            "call" -> (e["args"] as JsonArray).forEach { paths(it as JsonObject, out) }
            "list" -> (e["items"] as JsonArray).forEach { paths(it as JsonObject, out) }
        }
        return out
    }
}

File: packages/schema/src/expr.ts
/** Policy expressions: parser, evaluator, analysis. Spec: spec/01-schema.md §5. */
import type { Expr } from "./ir.ts";
import { TokenStream, tokenize } from "./lexer.ts";

export type BareRoot = "this" | "args";

export function parseExprText(text: string, bareRoot: BareRoot): Expr {
  const ts = new TokenStream(tokenize(text));
  const e = parseExpr(ts, bareRoot);
  if (!ts.at("eof")) throw ts.error("Unexpected token after expression");
  return e;
}

/** Parses an expression from the current position; stops at a token that cannot continue it. */
export function parseExpr(ts: TokenStream, bareRoot: BareRoot): Expr {
  return parseOr(ts, bareRoot);
}

function parseOr(ts: TokenStream, r: BareRoot): Expr {
  let l = parseAnd(ts, r);
  while (ts.accept("punct", "||")) l = { k: "bin", op: "||", l, r: parseAnd(ts, r) };
  return l;
}
function parseAnd(ts: TokenStream, r: BareRoot): Expr {
  let l = parseNot(ts, r);
  while (ts.accept("punct", "&&")) l = { k: "bin", op: "&&", l, r: parseNot(ts, r) };
  return l;
}
function parseNot(ts: TokenStream, r: BareRoot): Expr {
  if (ts.accept("punct", "!")) return { k: "not", e: parseNot(ts, r) };
  return parseCmp(ts, r);
}
const CMP_OPS = ["==", "!=", "<", "<=", ">", ">="] as const;
function parseCmp(ts: TokenStream, r: BareRoot): Expr {
  const l = parsePrimary(ts, r);
  const t = ts.peek();
  if (t.kind === "punct" && (CMP_OPS as readonly string[]).includes(t.value)) {
    ts.next();
    return { k: "bin", op: t.value as (typeof CMP_OPS)[number], l, r: parsePrimary(ts, r) };
  }
  if (t.kind === "name" && t.value === "in") {
    ts.next();
    return { k: "bin", op: "in", l, r: parsePrimary(ts, r) };
  }
  return l;
}
function parsePrimary(ts: TokenStream, r: BareRoot): Expr {
  const t = ts.peek();
  switch (t.kind) {
    case "string":
      ts.next();
      return { k: "lit", v: t.value };
    case "int":
    case "float":
      ts.next();
      return { k: "lit", v: t.num! };
    case "duration":
      ts.next();
      return { k: "lit", v: t.num! };
    case "punct":
      if (t.value === "(") {
        ts.next();
        const e = parseExpr(ts, r);
        ts.expectPunct(")");
        return e;
      }
      if (t.value === "[") {
        ts.next();
        const items: Expr[] = [];
        while (!ts.atPunct("]")) items.push(parseExpr(ts, r));
        ts.expectPunct("]");
        return { k: "list", items };
      }
      throw ts.error(`Unexpected ${JSON.stringify(t.value)} in expression`);
    case "name": {
      ts.next();
      if (t.value === "true") return { k: "lit", v: true };
      if (t.value === "false") return { k: "lit", v: false };
      if (t.value === "null") return { k: "lit", v: null };
      if (ts.atPunct("(")) {
        ts.next();
        const args: Expr[] = [];
        while (!ts.atPunct(")")) args.push(parseExpr(ts, r));
        ts.expectPunct(")");
        if (!BUILTIN_FNS.has(t.value)) throw ts.error(`Unknown function ${t.value}()`, t);
        return { k: "call", fn: t.value, args };
      }
      const path: string[] = [];
      while (ts.accept("punct", ".")) path.push(ts.expectName());
      if (t.value === "viewer" || t.value === "args" || t.value === "this") return { k: "path", root: t.value, path };
      return { k: "path", root: r, path: [t.value, ...path] };
    }
    default:
      throw ts.error("Expected expression");
  }
}

const BUILTIN_FNS = new Set(["has", "len", "now"]);

export interface ExprEnv {
  viewer: unknown;
  args: unknown;
  this: unknown;
  now?: () => number;
}

/** Total evaluation: missing paths are null, comparisons with null are false. */
export function evalExpr(e: Expr, env: ExprEnv): unknown {
  switch (e.k) {
    case "lit":
      return e.v;
    case "path": {
      let cur: unknown = env[e.root];
      // own members of objects only, as the JVM reads a JSON object: an array has none, and neither `length` nor
      // `constructor` is data, or a policy would decide differently on each runtime
      for (const p of e.path) {
        if (cur === null || typeof cur !== "object" || Array.isArray(cur) || !Object.hasOwn(cur, p)) return null;
        cur = (cur as Record<string, unknown>)[p];
      }
      return cur === undefined ? null : cur;
    }
    case "not":
      return !truthy(evalExpr(e.e, env));
    case "list":
      return e.items.map((i) => evalExpr(i, env));
    case "call": {
      const a = e.args.map((x) => evalExpr(x, env));
      switch (e.fn) {
        case "has":
          return Array.isArray(a[0]) && a[0].some((x) => looseEq(x, a[1]));
        case "len":
          return Array.isArray(a[0]) || typeof a[0] === "string" ? (a[0] as string | unknown[]).length : null;
        case "now":
          return env.now ? env.now() : Date.now();
        default:
          return null;
      }
    }
    case "bin": {
      if (e.op === "&&") return truthy(evalExpr(e.l, env)) && truthy(evalExpr(e.r, env));
      if (e.op === "||") return truthy(evalExpr(e.l, env)) || truthy(evalExpr(e.r, env));
      const l = evalExpr(e.l, env);
      const r = evalExpr(e.r, env);
      switch (e.op) {
        case "==":
          return looseEq(l, r);
        case "!=":
          return !looseEq(l, r);
        case "in":
          return Array.isArray(r) && r.some((x) => looseEq(l, x));
        default: {
          if (l === null || r === null || l === undefined || r === undefined) return false;
          const c = compareValues(l, r);
          if (e.op === "<") return c < 0;
          if (e.op === "<=") return c <= 0;
          if (e.op === ">") return c > 0;
          return c >= 0;
        }
      }
    }
  }
}

function truthy(v: unknown): boolean {
  return v !== null && v !== undefined && v !== false;
}

/** Raised when an expression compares values that cannot be ordered; a policy that errors fails closed. */
export class ExprError extends Error {}

/**
 * Equality that treats null/undefined as equal to each other only; a number equals numeric text of the same value
 * (Decimal and Long travel as text), and ids of any scalar type are equal by string.
 */
function looseEq(a: unknown, b: unknown): boolean {
  if (a === null || a === undefined) return b === null || b === undefined;
  if (b === null || b === undefined) return false;
  if (typeof a === "object" || typeof b === "object") return false;
  if (typeof a !== typeof b && (typeof a === "number" || typeof b === "number")) {
    const x = numericText(a);
    const y = numericText(b);
    if (x !== undefined && y !== undefined) return compareDecimal(x, y) === 0;
  }
  if (typeof a === typeof b) return a === b;
  return String(a) === String(b);
}

const DECIMAL_TEXT = /^-?\d+(\.\d+)?$/;

/** Order of two non-null values: numbers and numeric text compare exactly by value, other text by code point. */
function compareValues(a: unknown, b: unknown): number {
  const x = numericText(a);
  const y = numericText(b);
  if (x !== undefined && y !== undefined) return compareDecimal(x, y);
  if (typeof a === "string" && typeof b === "string") return a < b ? -1 : a > b ? 1 : 0;
  throw new ExprError(`Cannot order ${typeof a} and ${typeof b}`);
}

function numericText(v: unknown): string | undefined {
  if (typeof v === "number") {
    if (!Number.isFinite(v)) return undefined;
    const s = String(v);
    return /e/i.test(s) ? v.toLocaleString("en-US", { useGrouping: false, maximumFractionDigits: 20 }) : s;
  }
  return typeof v === "string" && DECIMAL_TEXT.test(v) ? v : undefined;
}

/** Exact comparison of decimal text, any length (no floating point, so values beyond 2^53 stay exact). */
function compareDecimal(a: string, b: string): number {
  const x = splitDecimal(a);
  const y = splitDecimal(b);
  if (x.neg !== y.neg) return x.neg ? -1 : 1;
  let c = x.int.length - y.int.length;
  if (c === 0) c = x.int < y.int ? -1 : x.int > y.int ? 1 : 0;
  if (c === 0) {
    const n = Math.max(x.frac.length, y.frac.length);
    const fx = x.frac.padEnd(n, "0");
    const fy = y.frac.padEnd(n, "0");
    c = fx < fy ? -1 : fx > fy ? 1 : 0;
  }
  return x.neg ? -Math.sign(c) : Math.sign(c);
}

function splitDecimal(t: string): { neg: boolean; int: string; frac: string } {
  const neg = t.startsWith("-");
  const [i = "0", f = ""] = (neg ? t.slice(1) : t).split(".");
  const int = i.replace(/^0+(?=\d)/, "");
  const frac = f.replace(/0+$/, "");
  return { neg: neg && !(int === "0" && frac === ""), int, frac };
}

export function exprToString(e: Expr): string {
  switch (e.k) {
    case "lit":
      return JSON.stringify(e.v);
    case "path":
      return [e.root, ...e.path].join(".");
    case "not":
      return `!${wrap(e.e)}`;
    case "list":
      return `[${e.items.map(exprToString).join(", ")}]`;
    case "call":
      return `${e.fn}(${e.args.map(exprToString).join(", ")})`;
    case "bin":
      return `${wrap(e.l)} ${e.op} ${wrap(e.r)}`;
  }
}
function wrap(e: Expr): string {
  return e.k === "bin" ? `(${exprToString(e)})` : exprToString(e);
}

/** Every path referenced by the expression. */
export function exprPaths(e: Expr): Array<{ root: "viewer" | "args" | "this"; path: string[] }> {
  const out: Array<{ root: "viewer" | "args" | "this"; path: string[] }> = [];
  const walk = (x: Expr): void => {
    switch (x.k) {
      case "path":
        out.push({ root: x.root, path: x.path });
        break;
      case "bin":
        walk(x.l);
        walk(x.r);
        break;
      case "not":
        walk(x.e);
        break;
      case "call":
        x.args.forEach(walk);
        break;
      case "list":
        x.items.forEach(walk);
        break;
      default:
    }
  };
  walk(e);
  return out;
}

export function referencesViewer(e: Expr): boolean {
  return exprPaths(e).some((p) => p.root === "viewer");
}

/**
 * Pushable = can be handed to a data source as a row filter: only viewer/args/literals and
 * top-level scalar fields of `this` (depth 1), no calls except has/len on viewer/args.
 */
export function isPushable(e: Expr): boolean {
  switch (e.k) {
    case "lit":
      return true;
    case "path":
      return e.root !== "this" || e.path.length === 1;
    case "not":
      return isPushable(e.e);
    case "bin":
      return isPushable(e.l) && isPushable(e.r);
    case "list":
      return e.items.every(isPushable);
    case "call":
      return e.fn !== "now" && e.args.every((a) => isPushable(a) && !(a.k === "path" && a.root === "this"));
  }
}
